### PR TITLE
Hyperbahn: add graceful shutdown support on sigterm

### PR DESCRIPTION
**Pull request by jcorbin**
r @Raynos

### CODE REVIEW

File: hyperbahn/clients/index.js [hunk 1]
```diff
@@ -139,6 +139,49 @@ function ApplicationClients(options) {
...
+
+    function onSigInt() {
+        if (!opts.tchannel.destroyed) {
+            self.logger.info('got SIGINT, closing tchannel', opts.tchannel.extendLogInfo({}));
+            opts.tchannel.close();
+        }
+    }
+
+    function isReqDrainExempt(req) {
+        // we only drain relay requests
```
Comment (anson627) [inline hyperbahn/clients/index.js]: why only relay?

File: hyperbahn/app.js [hunk 2]
```diff
@@ -66,7 +67,12 @@ function Application(config, opts) {
         rateLimiterBuckets: opts.rateLimiterBuckets
     });
     self.services = null;
+    self.logger = self.clients.logger;
     self.tchannel = self.clients.tchannel;
+    self.drainDeadlineTimeout = 30 * 1000;
```
Comment (Raynos) [inline hyperbahn/app.js]: nitpick: magic numbers should be NAMED_CONSTANTS at top of file.

File: hyperbahn/app.js [hunk 3]
```diff
@@ -113,6 +120,116 @@ Application.prototype.bootstrap = function bootstrap(cb) {
...
+        self.onSigInt();
+    }
+};
+
+Application.prototype.onSigTerm =
+function onSigTerm() {
+    var self = this;
+
+    if (self.tchannel.draining) {
+        self.logger.info('got additional SIGTERM while draining', self.tchannel.extendLogInfo({}));
```
Comment (Raynos) [inline hyperbahn/app.js]: nitpick: We should have a self.extendLogInfo() in future.

File: hyperbahn/app.js [hunk 4]
```diff
@@ -113,6 +120,116 @@ Application.prototype.bootstrap = function bootstrap(cb) {
...
+Application.prototype.startDrain =
+function startDrain() {
+    var self = this;
+
+    self.drainStart = self.tchannel.timers.now();
+    self.logger.info('got SIGTERM, draining application', self.tchannel.extendLogInfo({}));
+    self.tchannel.drain('shutting down due to SIGTERM',
+                        isReqDrainExempt, drainedThenClose);
+    self.drainDeadlineTimer = self.tchannel.timers.setTimeout(
+        self.deadlineTimedOut, self.drainDeadlineTimeout);
```
Comment (Raynos) [inline hyperbahn/app.js]: this does not work because `bind`

File: hyperbahn/app.js [hunk 5]
```diff
@@ -113,6 +120,116 @@ Application.prototype.bootstrap = function bootstrap(cb) {
...
+
+    if (!info) {
+        info = {};
+    }
+    var drainDuration = self.drainEnd - self.drainStart;
+    self.clients.statsd.timing('server.drain-time', drainDuration);
+    info.drainDurationMs = drainDuration;
+    info = self.tchannel.extendLogInfo(info);
+
+    self.logger.info(mess, info);
```
Comment (Raynos) [inline hyperbahn/app.js]: This level really should be different based on the message. Successful drain is info; timeout or sigints are warns

**Comment by Raynos** [comment]
Need to fix setTimeout and log level.

Otherwise lgtm

**Comment by Raynos** [comment]
lgtm; ship it.

**Pull request merged**